Yosemite/Yosemite/Tools/Stats/LeaderboardStatsConverter.swift

```diff
@@ -0,0 +1,82 @@
+import class Aztec.HTMLParser
+import struct Aztec.Element
+
+/// Helper functions to  convert a `Leaderboard` type into a `TopEearnerStat`
+///
+struct LeaderboardStatsConverter {
+
+    /// Returns the ids of the products associated with a top products leaderboar
+    ///
+    static func topProductsIDs(from topProducts: Leaderboard) -> [Int64] {
+        return topProducts.rows.compactMap {
+            Self.infeerProductID(fromHTMLString: $0.subject.display)
+        }
+    }
+
+    /// Indicates what products are missing from a leaderboard in an array of stored products
+    ///
+    static func missingProductsIDs(from topProducts: Leaderboard, in storedProducts: [Product]) -> [Int64] {
+
+        // Get the top products IDs
+        let topProductsIDs = Self.topProductsIDs(from: topProducts)
+
+        // Get the stored products IDs
+        let storedProductsIDs = storedProducts.map { $0.productID }
+
+        // Figure out which top products IDs are missing from store products IDs
+        let missingIDs = Set(topProductsIDs).subtracting(storedProductsIDs)
+        return Array(missingIDs)
+    }
+
+    /// Converts a the`leaderboard rows(top products)` into an array `TopEarnerStatsItem` using an array of stored products to match
+    ///
+    static func topEearnerStatItems(from topProducts: Leaderboard, using storedProducts: [Product]) -> [TopEarnerStatsItem] {
+        return topProducts.rows.compactMap {
+            Self.topEearnerStatItem(from: $0, using: storedProducts)
+        }
+    }
+}
+
+/// MARK: Private helpers
+private extension LeaderboardStatsConverter {
+
+    /// Infeers a product-id from an specific html string type
+    /// A valid html is an `a` tag with an `href` that includes the `product_id` in a query parameter named `products`
+    /// EG:  `<a href='https://store.com?products=9'>Product</a>`
+    ///
+    static func infeerProductID(fromHTMLString html: String) -> Int64? {
+
+        // Parse and extract the `products` parameter out the the html using `Aztec parser` and `URLComponents`
+        let parsed = HTMLParser().parse(html)
+        guard let a = parsed.firstChild(ofType: .a),
+            let href = a.attribute(ofType: .href)?.value.toString(),
+            let queryItems = URLComponents(string: href)?.queryItems,
+            let productItemValue = queryItems.first(where: { $0.name == "products"} )?.value else {
+                return nil
+        }
+
+        // Try to convert the `productID` to an `Int`
+        return Int64(productItemValue)
+    }
+
+    /// Converts a `leaderboard row(top product)` into a `TopEarnerStatsItem` using an array of stored products to match
+    ///
+    static func topEearnerStatItem(from topProduct: LeaderboardRow, using storedProducts: [Product]) -> TopEarnerStatsItem? {
+
+        // Match the product that corresponds to the leaderboard row(top product)
+        guard let productID = Self.infeerProductID(fromHTMLString: topProduct.subject.display),
```

Comment:
how's the performance of `infeerProductID` extracting the product ID from HTML? if it's not immediately fast, I was wondering if it's worth calling `func topProductsIDs` once in the beginning to get the product IDs and pass the product ID to this function

Reply:
I started from the baseline that parsing that HTML using Aztec parser was quite performant.
After running a quick `measure test` in Xcode I don't think we have much to worry about, especially because the product set is so small.

<img width="807" alt="Screen Shot 2020-07-23 at 3 40 05 PM" src="https://user-images.githubusercontent.com/562080/88336699-16d2e900-ccfb-11ea-8773-721c102c3d72.png">

Reply:
thanks for checking this Ernesto! sounds good 👍